sched/semaphore/sem_waitirq.c

```diff
@@ -72,31 +72,54 @@ void nxsem_wait_irq(FAR struct tcb_s *wtcb, int errcode)
 {
   FAR struct tcb_s *rtcb = this_task();
   FAR sem_t *sem = wtcb->waitobj;
+  bool mutex = NXSEM_IS_MUTEX(sem);
 
   /* It is possible that an interrupt/context switch beat us to the punch
    * and already changed the task's state.
    */
 
-  DEBUGASSERT(sem != NULL && atomic_read(NXSEM_COUNT(sem)) < 0);
+  DEBUGASSERT(sem != NULL);
+  DEBUGASSERT((mutex || atomic_read(NXSEM_COUNT(sem)) < 0) &&
+              (!mutex || NXSEM_MBLOCKING(atomic_read(NXSEM_MHOLDER(sem)))));
```

Comment:
let's spit to two assert like other similar places, which also make the assertion trigger condition more clear.